### PR TITLE
TRD tracking as a DPL workflow

### PR DESCRIPTION
**Pull request by martenole**
Hello @shahor02 , @davidrohr  @tdietel  and @bazinski 

I created a first version of a TRD tracking workflow (currently based on ITS-TPC matched tracks as input). At the moment it has basically the same functionality as the macro I used before `run_trd_tracking.C`. Here is a list of what I would like to achieve in the end:

- The `TRDGlobalTrackingSpec` should get an option whether to use ITS-TPC matched tracks or TPC only tracks as input. The functionality to use TPC only tracks of course needs to be added to `GPUTRDTracker` first.
- At some point the TRD tracking shall receive calibrated tracklets which are available after the standalone TRD tracking. But since the development has just started I would like to take the "raw" online tracklets as a first step and start with those. I will simply transform them into the Run2 format and plug them into the tracker. This will be done as soon as the tracklets give reasonable results.
- I wonder what the output of the TRD tracking whould be for the synchronous reconstruction. For the TPC interpolation we will not need the tracks, but only the indices of the global tracks, which could be matched successfully to the TRD and for those tracks the up to 6 space points. For the TPC interpolation these space points will then be used for a refit from the outer radius. Maybe I need to add this to the `GPUTRDTracker` class.
- MC is completely disabled at the moment.

Open questions:
- In the `run` method of `TRDGlobalTrackingSpec.cxx` I cannot use `gsl::span` as input for the TRD data, but only `std::vector` since the data is not trivially copyable. I do not know why.

I wanted to ping also Jason. Does somebody know his github name?

Cheers,
Ole

### CODE REVIEW

**Comment by shahor02** [comment]
Hi @martenole 
below is my opinion.

>     * The `TRDGlobalTrackingSpec` should get an option whether to use ITS-TPC matched tracks or TPC only tracks as input. The functionality to use TPC only tracks of course needs to be added to `GPUTRDTracker` first.

Since the effective difference between the ITS-TPC and TPC tracks is the strictness of the time constraint, one can probably have the same code with templated input of the track. It just need to be ready to use an asymmetric time constraint (e.g. for TPC tracks the time bracket is given by ``getTime0() - getDeltaTBwd()`` : ``getTime0() - getDeltaTFwd()`` and the units (for ITS-TPC tracks it is \mus, for TPC tracks: timebins. This can be handled by an appropriate scaling parameter.

I think the option which input to use should be an option not of the device spec but of the workflow: the output messages should have different DataDescriptions and I think the device name also should reflect what is used (so that we can have both as "different" devices in the same piped workflow): this means that the option should act on the level of https://github.com/AliceO2Group/AliceO2/blob/a6cb8232a4dde44bdfde06999fd969a108f03271/Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx#L176
 
>     * I wonder what the output of the TRD tracking whould be for the synchronous reconstruction. For the TPC interpolation we will not need the tracks, but only the indices of the global tracks, which could be matched successfully to the TRD and for those tracks the up to 6 space points. For the TPC interpolation these space points will then be used for a refit from the outer radius. Maybe I need to add this to the `GPUTRDTracker` class.
> 
I think this should be a 
(1) vector of "TRD tracks" with updated track params (I would consider both outward and inward params), the index of the source TPC or TPC-ITS track and the start + number of tracklet indices matched to this track, 
(2) a vector of used tracklet indices

> 
> Open questions:
> 
>     * In the `run` method of `TRDGlobalTrackingSpec.cxx` I cannot use `gsl::span` as input for the TRD data, but only `std::vector` since the data is not trivially copyable. I do not know why.
> 
This is because the Tracklet64 has non-trivial copy constructor. You have to impose the messageability using this trick:
https://github.com/AliceO2Group/AliceO2/blob/1b8bea430dd6c8d5243691bca49e93dd1e3a30ba/DataFormats/Detectors/ITSMFT/common/include/DataFormatsITSMFT/Cluster.h#L201-L208

**Comment by bazinski** [comment]
Hi @martenole and @shahor02 
Open questions:
> > ```
> > * In the `run` method of `TRDGlobalTrackingSpec.cxx` I cannot use `gsl::span` as input for the TRD data, but only `std::vector` since the data is not trivially copyable. I do not know why.
> > ```
> 
> This is because the Tracklet64 has non-trivial copy constructor. You have to impose the messageability using this trick:
> https://github.com/AliceO2Group/AliceO2/blob/1b8bea430dd6c8d5243691bca49e93dd1e3a30ba/DataFormats/Detectors/ITSMFT/common/include/DataFormatsITSMFT/Cluster.h#L201-L208

Would it not be better to simply make the copy constructor trivial? it only has a uint64_t as a member so trivial memberwise copy should be fine? I dont remember why I put in a trival, non-trivial copy constructor, I assume its a carry over from previous implementations while testing, and can therefore go. Will change shortly.

As for Jason...
@jbarrella meet @martenole ;-)

**Comment by shahor02** [comment]
@bazinski yes, relying on the default copy c-tor would be more natural. The ``is_messageable`` is a brute-force solution

**Comment by martenole** [comment]
@bazinski I added the change in Tracklet64 in a commit here

File: Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx
Comment (davidrohr) [inline Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx]: What is the meaning of a TRD track with 6 missing layers if TRD has only 6 layers?

File: Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx
Comment (davidrohr) [inline Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx]: I would not use debug level 5 for production code.

File: Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx
Comment (davidrohr) [inline Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx:66]: If you set trivial objects, you can as well just use `GPUReconstruction::SetSettings(float solenoidBz)`, and btw you should set the solenoid field at least.

File: Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx
Comment (davidrohr) [inline Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx:85]: I think in the long run, we should move these parameters that will be relevant to `GPUSettingsRec` and then use `--configKeyValues` to set them, as it is already done in the TPC workflow. But I would do this in an independent PR, and first try to get the workflow going.

**Review by davidrohr — COMMENTED** [review]
Looks in principle OK to me, few comments inline.
In general, we should think at some point about how we integrate this with the GPU workflow, when we want to bring the full chain on the GPU. Yet, this misses TPC-ITS matching (for the case of using the TRD tracker on TPC-ITS matched tracks), so I think this is not a topic for now. I'd go ahead with this workflow more or less as is after some polishing. Anyway, it is not that much code, and also not much code duplication, so for now it seems fine to me.

File: Detectors/TRD/workflow/src/TRDTrackletReaderSpec.cxx
Comment (jbarrella) [inline Detectors/TRD/workflow/src/TRDTrackletReaderSpec.cxx]: I borrowed this TrackletReader device for my own workflow and noticed that you use a different name for the trigger record than what Sean uses in the TRAPsim. In the TRAPsim, it is called `TRKTRGRD`. Would it make more sense to use the same name here?

File: Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx
Comment (martenole) [reply Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx]: If this would be set to 5, then the track following would be stopped after the fifth layer in case no tracklet can be matched to the track up to that point. This could be tuned. I need to check how often a track with a tracklet only in the 6th layer is found.
In general these settings should be settable from the command line I guess.

File: Detectors/TRD/workflow/src/TRDTrackletReaderSpec.cxx
Comment (martenole) [reply Detectors/TRD/workflow/src/TRDTrackletReaderSpec.cxx]: Good point, the naming should be consistent. I have changed this.

File: Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx
Comment (davidrohr) [reply Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx]: ok, but then I the variable name seems to be misleading. I have interpreted SetMaxMissingLayers(5) as:
if we arrirve at layer 6 without a match, we have 5 missing layers, which is still OK, so go on :)

**Comment by martenole** [comment]
@davidrohr Thanks for your comments. I still need to implement them, but at the moment I have troubles running the workflow due to the TRD geometry. Apparently it is not initialized correctly within the workflow (although when the tracking is run from the macro `run_trd_tracker.C` it works).
Is there a command line argument for o2 workflows to limit the processing to 1 thread? This would make it easier to debug.

I see errors of the form:
[209077:trd-globaltracking]: [14:23:02][ERROR] Found TRD tracklets in chamber 254 which is not included in the geometry
[209077:trd-globaltracking]: [14:23:02][ERROR] Space points for at least one chamber could not be calculated
[209077:trd-globaltracking]: [14:23:02][ERROR] Found TRD tracklets in chamber 251 which is not included in the geometry
[209077:trd-globaltracking]: [14:23:02][ERROR] Space points for at least one chamber could not be calculated
[209077:trd-globaltracking]: [14:23:02][ERROR] Found TRD tracklets in chamber 65 which is not included in the geometry
[209077:trd-globaltracking]: [14:23:02][ERROR] Space points for at least one chamber could not be calculated
[209077:trd-globaltracking]: [14:23:02][ERROR] Found TRD tracklets in chamber 167 which is not included in the geometry
[209077:trd-globaltracking]: [14:23:02][ERROR] Space points for at least one chamber could not be calculated
[209077:trd-globaltracking]: [14:23:02][ERROR] Found TRD tracklets in chamber 10 which is not included in the geometry
[209077:trd-globaltracking]: [14:23:02][ERROR] Space points for at least one chamber could not be calculated
[209077:trd-globaltracking]: [14:23:02][ERROR] Found TRD tracklets in chamber 20 which is not included in the geometry
[209077:trd-globaltracking]: [14:23:02][ERROR] Space points for at least one chamber could not be calculated
[209077:trd-globaltracking]: [14:23:02][ERROR] Found TRD tracklets in chamber 171 which is not included in the geometry
[209077:trd-globaltracking]: [14:23:02][ERROR] Space points for at least one chamber could not be calculated
[209077:trd-globaltracking]: [14:23:02][ERROR] Found TRD tracklets in chamber 30 which is not included in the geometry
[209077:trd-globaltracking]: [14:23:02][ERROR] Space points for at least one chamber could not be calculated



And the processing should be aborted in case only one matrix is not found which is expected to be there. And all these chambers which are reported missing should actually be there. Maybe I run into thread-safety issues here, this I would like to check.

File: Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx
Comment (martenole) [reply Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx]: Maybe `SetStopTrkFollowingAfterNMissingLayers(6) // never abandon tracks before all layers have been checked` is more clear? :)

**Comment by davidrohr** [comment]
you can set the env variable OMP_NUM_THREADS=1 to limit omp to 1 thread.

**Comment by martenole** [comment]
Thanks, but I still see the error message 8 times, although it should appear only once. Within the terminal I simply type
OMP_NUM_THREADS=1
o2-trd-global-tracking  -b
Is there another flag I need to set? Or a CLI option for the DPL workflow driver? I could not find anything with o2-trd-global-tracking  --help=full

**Comment by davidrohr** [comment]
The DPL workflow should not change anything for the OMP threading.
And OMP_NUM_THREADS=1 works as intended for the TPC workflow.
Can you try if it works with your macro?

**Comment by martenole** [comment]
With the macro I never get the problems, because only the track following is parallelized with openmp. The function in question which throws the error `CalculateSpacePoints()` is called only from a single instance.
When I run the trd-globaltracking workflow then it seems that everything is parallelized, independent of the env setting OMP_NUM_THREADS=1.

**Comment by davidrohr** [comment]
ok, but then it is not openmp parallelization.
Without checking your code: perhaps there is some parallelization over the TRD sectors in the workflow?

**Comment by martenole** [comment]
No, the workflow is fairly simple and I don't add parallelization.
In the init file which is created I see the following lines:
...
[GPU_proc]
...
ompThreads=-1
...
nStreams=8
...

Can it be the nStreams setting? Do you know how I could change this?

**Comment by davidrohr** [comment]
the nStreams affects only processing on the GPU, shouldn't matter at all.
Can you run with the macro and compare the settings? They should be the same. If they are the same indeed, it should tell you that nothing is wrong with them.

**Comment by martenole** [comment]
I debugged a bit further and noticed it is not an issue with parallelization... The omp flag works fine. I called CalculateSpacePoints in a loop over all interactions without a break statement and that is why I saw multiple error messages. Unfortunately, by chance there were 8 interactions with TRD tracklets, the same number as the number of threads I have on my computer leading to confusion :(

Somehow the memory layout seems to get messed up within the workflow between a call to `TRDGlobalTracking::init` and `TRDGlobalTracking::run`. During init the TRD geometry which is loaded into the tracker looks fine. But then in the run method when I try to access the flat geometry the entries are nonsense. I have no idea why.
With the latest commit one can try this by running `o2-trd-global-tracking`. During `init` the geometry is printed twice and the entries are fine. In the subsequent `run` method the entries of `mMatrixIndirection` of `GeometryFlat` are random.

**Comment by davidrohr** [comment]
@martenole : I believe your problem is that the objects in you init function go out of scope and are thus destroyed. Hence you read garbage when accessing them during the processing. You can have a look at the CATrackerSpec for the TPC by Matthias which uses some trick with shared ptrs and lambda captures to keep the objects alive as long as the lambda lives.

**Comment by martenole** [comment]
Thanks a lot @davidrohr this was exactly the case. For now I did not implement a very smart solution, I simply added the flat geometry as a member to the tracker spec. But this looks good now :)

File: Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx
Comment (martenole) [reply Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx:66]: I use now the function as you suggested and set the solenoid field. That is important for the GPU reconstruction, but in case of the CPU version when I use the o2::base::Propagator within the TRD tracking I think this does not matter because of the `initFieldFromGRP` call above. But anyhow, as soon as we switch to the GPU data types this is of course necessary so I added it.

File: Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx
Comment (martenole) [reply Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx]: I am using now the default values as you suggested

Comment (martenole) [reply Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx]: Nevermind, simply the default values won't work. I need to set the GPURecoStepConfiguration at least, because otherwise I get the following errors:
[205400:trd-globaltracking]: [14:33:32][ERROR] Invalid input, TPC Clusterizer needs TPC raw input
[205400:trd-globaltracking]: [14:33:32][FATAL] GPUReconstruction could not be initialized

Comment (davidrohr) [reply Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx]: @martenole : Indeed you obviously need to set the recosteps. Can you retry with #4813 which has the correct setter for it :)

Comment (martenole) [reply Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx]: Thanks for adding that setting, but I would keep it as it is for now, so that I can easily adjust debug level or other settings during the development. I guess as soon as the tracklets are ready I think this will be needed..

**Comment by martenole** [comment]
Did a fresh rebase to resolve a merge conflict with dev. This PR should (I think) be ready to be merged. Please let me know in case you have additional comments.

**Review by davidrohr — DISMISSED** [review]
looks ok for me, further improvements can come later.
What do I need to do to run it? Do I have to prepare the trd tracklets somehow, or is the digitizer workflow all that is needed?

**Comment by martenole** [comment]
I do the full chain such that I have ITS-TPC matched tracks and the tracklets. The steps I do are:

- generate 10 pp events with pythia8 as generator
o2-sim -n 10 -g pythia8 --skipModules ZDC
- run digitization for all detectors
o2-sim-digitizer-workflow -b
- run ITS reconstruction
o2-its-reco-workflow -b
- run TPC reconstruction
o2-tpc-reco-workflow -b --tpc-digit-reader '--infile tpcdigits.root' --input-type digits --output-type clusters,tracks --tpc-track-writer "--treename events --track-branch-name Tracks --trackmc-branch-name TracksMCTruth"
- run ITS-TPC matching
o2-tpcits-match-workflow -b  --tpc-track-reader tpctracks.root --tpc-native-cluster-reader "--infile tpc-native-clusters.root" 
- run TRD TRAP Sim
o2-trd-trap-sim -b
- finally run TRD tracking
o2-trd-global-tracking -b

**Comment by martenole** [comment]
Hi @tdietel 
is this workflow also OK for you? I think it still requires your approval as code owner. I am not sure why the automatic tests take so long at the moment, but it seems this is a general problem for all current PRs.
Cheers,
Ole

**Comment by bazinski** [comment]
@martenole just a FYI, you can run the digitizer with --enable-trd-trapsim  and you dont have to run the trapsim seperately.

**Review by bazinski — DISMISSED** [review]

**Review by tdietel — DISMISSED** [review]
I did not manage to go into the details, but I see that many points were already discussed. Based on these discussion, I give my approval, too.

**Comment by davidrohr** [comment]
So, I finally tried it today and it seems to work correctly on pp data. One remark I have: I think there is quite a lot of console output, per every track if I understand correctly. I think for production code this should be disabled.

I am also trying with Pb-Pb events but the trap simulator is already running for 10 minutes (I have 10 events). I this expected, seems excessively long to me. Have you tried with Pb-Pb events at all?

**Comment by bazinski** [comment]
Hi @davidrohr 10 minutes sounds rather long, can you send me the digits file to take a look? It should be doing about 7k tracklets per second so in 10 minutes at that rate is one hell of a lot of tracklets (my desktop performance), sounds more likely something is going wrong.

**Comment by davidrohr** [comment]
@bazinski : Please find the digit file here: https://qon.jwdt.org/nmls/tmp/sim.tar.gz
I also see a message
```
[55663:TRAP]: [13:26:07][INFO] TRD Trap Simulator Device running over incoming message
```
Not sure what it means, but I also see it with the pp data set, where it does not get stuck.

**Comment by bazinski** [comment]
@davidrohr That message is just to say the trapsimulator has started, maybe slightly overly verbose ;-)

**Comment by martenole** [comment]
@davidrohr I have cleaned up the debug output for default settings.

I have not tried PbPb so far. I wanted to do it now but I get the following errors:
Command: `o2-sim -n 10 -g hijing --skipModules ZDC`
Output:
[INFO] Running with 4 sim workers 
[INFO] CREATING SIM SHARED MEM SEGMENT FOR 4 WORKERS
Spawning particle server on PID 29766; Redirect output to o2sim_serverlog
Spawning sim worker 0 on PID 29768; Redirect output to o2sim_workerlog0
Spawning hit merger on PID 29769; Redirect output to o2sim_mergerlog
[INFO] Process 29766 EXITED WITH CODE 0 SIGNALED 1 SIGNAL 6
[FATAL] ABORTING DUE TO ABORT IN COMPONENT
[INFO] Process 29768 EXITED WITH CODE 0 SIGNALED 1 SIGNAL 6
[FATAL] ABORTING DUE TO ABORT IN COMPONENT
[INFO] Merger process 29769 returned
[INFO] Simulation process took 25.2642 s
Error in <TFile::ReadBuffer>: error reading all requested bytes from file o2sim_Kine.root, got 226 of 300
Error in <TFile::Init>: o2sim_Kine.root failed to read the file type data.

Do I need to run with externally provided kinematics as Ruben described in the mail he send to wp13 in January?
o2-sim -n $nev --skipModules ZDC  -g extkin --extKinFile $wd/kinem/Kinematics_pbpb_200.root 2>&1 | tee sim_hijing.log

**Comment by davidrohr** [comment]
Thx

@martenole : Hijing is not supported yet, you'd need to run with external kinematics from AliRoot.
But you can use pythia8hi instead, which is a new pythia Pb-Pb generator. (just use pythia8hi instead of pythia8).

**Comment by bazinski** [comment]
rebuilt head of dev, and timing o2-trd-trap-sim :
`o2-trd-trap-sim -b  63.79s user 3.14s system 262% cpu 25.533 total`

Running on the trddigits your linked above? Its doing it at about 6k/s on my desktop.

**Comment by davidrohr** [comment]
> rebuilt head of dev, and timing o2-trd-trap-sim :
> `o2-trd-trap-sim -b 63.79s user 3.14s system 262% cpu 25.533 total`
> 
> Running on the trddigits your linked above. Its doing it at about 6k/s on my desktop.

Interesting, this is with the digits file I provided?

**Comment by bazinski** [comment]
oui oui

**Comment by davidrohr** [comment]
ok, that is very strange, I retried with some different simulations and get the same result. With pp data it seems to work.
And from the debugger, it seems to be stuck in some sorting. I'll try to have a closer look later.
@martenole : Could you try and let me know whether it works for you?

**Comment by bazinski** [comment]
will try replicate in other places, i am on ubuntu 18.04.

**Comment by martenole** [comment]
I am also stuck for 10 minutes now with the trap sim. Not with your digits file David, but with a simulation I did myself with pythia8hi (10 events as well)

**Comment by bazinski** [comment]
do i need to reinstall my machine :-(

**Comment by bazinski** [comment]
Well this issue does not appear to be related to the contents of this PR, maybe move it to a jira ticket or something else?

**Review by davidrohr — APPROVED** [review]

**Comment by davidrohr** [comment]
Indeed, this one is OK, can be merged once the CI passes. I'll see what goes wrong in the trap sim and open a JIRA then.

**Pull request merged**
